src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -66,73 +97,23 @@ where
     }
 }
 
-fn copy_to<G>(
-    sinked_collection: Collection<G, Row, Diff>,
-    err_collection: Collection<G, DataflowError, Diff>,
-    sink_id: GlobalId,
-    up_to: Antichain<G::Timestamp>,
-    response_protocol_handle: Rc<RefCell<Option<ResponseProtocol>>>,
-) where
-    G: Scope<Timestamp = Timestamp>,
-{
-    let name = format!("s3-oneshot-{}", sink_id);
-    let mut op = OperatorBuilder::new(name, sinked_collection.scope());
-    let mut ok_input = op.new_input(&sinked_collection.inner, Pipeline);
-    let mut err_input = op.new_input(&err_collection.inner, Pipeline);
-
-    op.build(|_cap| {
-        let mut row_count: u64 = 0;
-        let mut finished = false;
-        let mut ok_buf = Default::default();
-
-        move |frontiers| {
-            if finished {
-                // Drain the inputs, to avoid the operator being constantly rescheduled
-                ok_input.for_each(|_, _| {});
-                err_input.for_each(|_, _| {});
-                return;
-            }
-
-            let mut frontier = Antichain::new();
-            for input_frontier in frontiers {
-                frontier.extend(input_frontier.frontier().iter().copied());
-            }
-
-            ok_input.for_each(|_, data| {
-                data.swap(&mut ok_buf);
-                let count: u64 = ok_buf.drain(..).len().try_into().expect("usize into u64");
-                // TODO (mouli): write these to s3 in subsequent PRs
-                row_count = row_count + count;
-            });
-            // TODO (mouli): handle errors
-            if PartialOrder::less_equal(&up_to, &frontier) {
-                finished = true;
-                // We are done, send the final count.
-                if let Some(response_protocol) = response_protocol_handle.borrow_mut().deref_mut() {
-                    response_protocol.send_final_count(row_count);
-                }
-            }
-        }
-    });
-}
-
 /// A type that guides the transmission of number of rows back to the coordinator.
 struct ResponseProtocol {
     pub sink_id: GlobalId,
     pub response_buffer: Option<Rc<RefCell<Vec<(GlobalId, CopyToResponse)>>>>,
 }
 
 impl ResponseProtocol {
-    fn send_final_count(&mut self, rows: u64) {
-        let buffer = self.response_buffer.as_mut().expect("Copy response buffer");
-
-        buffer
-            .borrow_mut()
-            .push((self.sink_id, CopyToResponse::RowCount(rows)));
-
+    // This method should only be called once otherwise this will panic.
```

Comment:
This wasn't the case before, is there a reason we have this restriction now?

Reply:
Actually it was the same before. Just that it was getting the unwrapped value from `response_buffer.as_mut().expect` and then set the `response_buffer` to `None` = equivalent to the `response_buffer.take()` here.

And we do want it to be strictly one message back per worker, otherwise it can mess up the way the final response is accumulated across all the workers for the replica.